components/num-util/src/fixed_decimal.rs

```diff
@@ -0,0 +1,517 @@
+    /// assert_eq!(Error::Limit, dec1.adjust_magnitude(std::i16::MAX).unwrap_err());
+    /// ```
+    Limit,
+}
+
+/// A struct containing decimal digits with efficient iteration and manipulation by magnitude
+/// (power of 10). Supports a mantissa of non-zero digits and a number of leading and trailing
+/// zeros, used for formatting and plural selection.
+///
+/// You can create a FixedDecimal from an unsigned integer or an isize.
+///
+/// # Examples
+///
+/// ## From an isize
+///
+/// ```
+/// use icu_num_util::FixedDecimal;
+///
+/// let dec1 = FixedDecimal::from(250);
+/// assert_eq!("250", dec1.to_string());
+/// ```
+///
+/// ## From a BigUint
+///
+/// ```
+/// use icu_num_util::FixedDecimal;
+/// use icu_num_util::fixed_decimal::UnsignedWrapper;
+/// use num::BigUint;
+/// use std::convert::TryFrom;
+///
+/// let input = BigUint::parse_bytes(b"98765432109876543210", 10).unwrap();
+/// let dec2 = FixedDecimal::try_from(UnsignedWrapper(input)).unwrap();
+/// assert_eq!("98765432109876543210", dec2.to_string());
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct FixedDecimal {
+    /// List of digits; digits[0] is the most significant.
+    ///
+    /// Invariants:
+    /// - Must not include leading or trailing zeros
+    /// - Length must not exceed (magnitude - lower_magnitude + 1)
+    // TODO: Consider using a nibble array
+    digits: SmallVec<[u8; 8]>,
+
+    /// Power of 10 of digits[0].
+    ///
+    /// Invariants:
+    /// - <= upper_magnitude
+    /// - >= lower_magnitude
+    magnitude: i16,
+
+    /// Power of 10 of the most significant digit, which may be zero.
+    ///
+    /// Invariants:
+    /// - >= 0
+    /// - >= magnitude
+    upper_magnitude: i16,
+
+    /// Power of 10 of the least significant digit, which may be zero.
+    ///
+    /// Invariants:
+    /// - <= 0
+    /// - <= magnitude
+    lower_magnitude: i16,
+
+    /// Whether the number is negative. Negative zero is supported.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu_num_util::FixedDecimal;
+    ///
+    /// let mut dec: FixedDecimal = Default::default();
+    /// dec.is_negative = true;
+    /// assert_eq!("-0", dec.to_string());
+    /// ```
+    pub is_negative: bool,
+}
+
+impl Default for FixedDecimal {
+    /// Returns a FixedDecimal representing zero.
+    fn default() -> Self {
+        FixedDecimal {
+            digits: SmallVec::new(),
+            magnitude: 0,
+            upper_magnitude: 0,
+            lower_magnitude: 0,
+            is_negative: false,
+        }
+    }
+}
+
+impl From<isize> for FixedDecimal {
+    fn from(value: isize) -> Self {
+        let value_abs = if value == std::isize::MIN {
+            // isize::MIN.abs() is too big for isize by 1
+            std::isize::MAX as usize + 1
+        } else {
+            value.abs() as usize
+        };
+        let mut result = FixedDecimal::from_ascending(UintIterator::from(value_abs)).unwrap();
+        result.is_negative = value < 0;
```

Comment:
this will set `is_negative = false` for `-0`.

Reply:
Good point, but can `-0` be represented by an `isize`?

Reply:
good point :)